src/Bus.cc

```diff
@@ -933,4 +934,101 @@ namespace kickcat
             link_.addDatagram(Command::FPRD, createAddress(slave.address, reg::ERROR_COUNTERS), nullptr, sizeof(ErrorCounters), process, error);
         }
     }
+
+    void Bus::getTopology()
+    {
+        uint16_t lastSeen = slaves_.at(0).address;
+        std::stack<uint16_t> branches;
+        for (auto& slave : slaves_) 
+        {   
+            int openPorts = slave.countOpenPorts();
+
+            if (openPorts == 1)
+            {
+                topology_[slave.address] = lastSeen;
+                lastSeen = slave.address;
+                if (!branches.empty())
+                {
+                    lastSeen = branches.top();
+                    branches.pop();
+                }
+            }
+            else if (openPorts == 2)
+            {
+                topology_[slave.address] = lastSeen;
+                lastSeen = slave.address;
+            }
+            else 
+            {
+                topology_[slave.address] = lastSeen;
+                lastSeen = slave.address;
+                for (int i = 2; i < openPorts/2; i++)
+                {
+                    branches.push(slave.address);
+                }
+            }
+        }
+    }
+
+    void Bus::printBasicTopology()
+    {  
+        printf( "\n -*-*-*-*- Topology -*-*-*-*-\n" );
+        for (auto& slave : slaves_)
+        {
+            if (slave.address != topology_[slave.address])
+            {
+                printf( "Slave %04x parent : slave %04x \n", slave.address, topology_[slave.address]);
+            }
+            else
+            {
+                printf( "Slave %04x parent : master \n", slave.address);
+            }
+        }
+    }
+
+    void Bus::sendGetRegister(Slave& slave, uint16_t reg)
```

Comment:
Are all registers guaranteed to be 16bits large? If no a template may be a good option.